### PR TITLE
feat: AKS user-assigned identity for control plane and kubelet

### LINKED ISSUE #3239: Support User Identity in CAPZ managed AKS Cluster
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Currently, only system managed identity is supported on CAPZ managed AKS cluster for **control plane identity**. Based on your current use case, it would be great to have user managed identity support in AKS cluster.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

Code path where we hard code identity to system managed identity: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/managedclusters/spec.go#L273

**Environment:**

- cluster-api-provider-azure version:  1.7.2
- Kubernetes version: (use `kubectl version`):  
- OS (e.g. from `/etc/os-release`):

### LINKED ISSUE #3261: Support Kubelet User Managed Identity on AKS VMSS pool.
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Currently, we reply on an auto-assigned **kubelet user identity** which is assigned via during AKS cluster creation. AKS provides capability to configure BYO user identity in the kubelet identity field. This enables user to define permissions before the cluster is created and assign the identity during cluster creation.

Link: https://learn.microsoft.com/en-us/azure/aks/use-managed-identity#use-a-pre-created-kubelet-managed-identity

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
Just to clarify, there are two types of identity supported in AKS. first is the control plane identity and other is the VMSS pool identity. The new feature ask is for the later one where we require pods running on AKS to have access to ARM resources.
Correspond to [this](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/containerservice/armcontainerservice/models.go#L1530).)

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR introduces the option to pass in a pre-existing user-assigned identity for AKS cluster identity. See the top two rows here:

- https://learn.microsoft.com/en-us/azure/aks/use-managed-identity#summary-of-managed-identities

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #3239
Fixes #3261

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
AKS user-assigned identity for control plane and kubelet
```

### CODE REVIEW

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (alexeldeib) [inline api/v1beta1/azuremanagedcontrolplane_types.go:56]: language nit: I believe we prefer control plane these days

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (alexeldeib) [inline api/v1beta1/azuremanagedcontrolplane_webhook.go:689]: have you tested this? I'm guessing you're following https://learn.microsoft.com/en-us/azure/aks/use-managed-identity ?

principal ID and client ID are _outputs_ there I thought, iirc you only need the UAMI ARM ID to pass to AKS. could be wrong.

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (alexeldeib) [inline api/v1beta1/azuremanagedcontrolplane_webhook.go:699]: I don't even think you should expose these fields except in status. but see above. 

if I'm mistaken, this makes sense to me.

**Review by alexeldeib — COMMENTED** [review]
nice work :) not sure we actually need principal and client ID. I believe AKS will resolve these internally and emit them as outputs, and would ignore inputs. Worth double checking though.

I checked some of my dev HTTP requests in postman and they only use ARM ID, but that stuff circumvents ARM, so I'm not sure it's correct, especially for system identity (which is tricky in dev env)

**Comment by CecileRobertMichon** [comment]
@jackfrancis are you still working on this one?

File: azure/services/managedclusters/spec.go
Comment (jackfrancis) [inline azure/services/managedclusters/spec.go:410]: @alexeldeib thanks for the review (and sorry for the delayed response)

Are you suggesting that these are the wrong inputs to pass into the AKS API?

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (luthermonson) [inline api/v1beta1/azuremanagedcontrolplane_types.go:52]: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/types.go#L520-L531 can we just reuse these existing consts?

**Comment by luthermonson** [comment]
> nice work :) not sure we actually need principal and client ID. I believe AKS will resolve these internally and emit them as outputs, and would ignore inputs. Worth double checking though.
> 
> I checked some of my dev HTTP requests in postman and they only use ARM ID, but that stuff circumvents ARM, so I'm not sure it's correct, especially for system identity (which is tricky in dev env)

Comparing to how the azurerm terraform provider solves this exact problem I believe you are correct and this feature is also designed to be receive multiple user assigned IDs. Please check out the following code:

https://github.com/hashicorp/terraform-provider-azurerm/blob/8bc0e5637346eeea808f5b63b97eed69835005b5/internal/services/containers/kubernetes_cluster_resource.go#L3265-L3283

I think a better API would be to reuse the idea in AzureMachinePools which is the (UserAssignedIdentities)[https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1beta1/azuremachinepool_types.go#L144C6-L149] list and do a loop that looks like the one in the tf provider to add empty `UserAssignedIdentityDetails` with the ID as the map key.

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:696]: It's a bit of a mystery!

E0619 20:02:48.596549       1 controller.go:329]  "msg"="Reconciler error" "error"="error creating AzureManagedControlPlane capz-e2e-rwuh71/capz-e2e-rwuh71-aks: failed to reconcile AzureManagedControlPlane service managedcluster: failed to create resource capz-e2e-rwuh71-aks/capz-e2e-rwuh71-aks (service: managedcluster): Code=\"InvalidIdentityValues\" Message=\"Invalid value for the identities '/subscriptions/<redacted>/resourcegroups/capz-ci/providers/Microsoft.ManagedIdentity/userAssignedIdentities/umd-test'. The 'UserAssignedIdentities' property keys should only be empty null during PATCH requests.\"" "AzureManagedControlPlane"={"name":"capz-e2e-rwuh71-aks","namespace":"capz-e2e-rwuh71"} "controller"="azuremanagedcontrolplane" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedControlPlane" "name"="capz-e2e-rwuh71-aks" "namespace"="capz-e2e-rwuh71" "reconcileID"="b8fe5ded-9132-42cc-89ea-ae1621f4dbcf"

The above is in response to only passing in the URI as the key as one does here:

- https://github.com/Azure/azure-sdk-for-go/blob/v67.0.0/services/containerservice/mgmt/2022-03-01/containerservice/models.go#L1770-L1780

Your assertion that the principalID and clientID are output does check out with the type description here:

- https://github.com/Azure/azure-sdk-for-go/blob/v67.0.0/services/containerservice/mgmt/2022-03-01/containerservice/models.go#L1794-L1800

I'll try hacking around with a pointer to an empty `ManagedClusterIdentityUserAssignedIdentitiesValue` literal as the value instead of `nil` and see if it still complains about PATCH (or complains about anything), stand by....

Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:696]: yep that worked!

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:52]: I'd prefer to keep these strongly typed and distinct, but I have inherited the actual string const values from the vm consts

**Comment by jackfrancis** [comment]
> > nice work :) not sure we actually need principal and client ID. I believe AKS will resolve these internally and emit them as outputs, and would ignore inputs. Worth double checking though.
> > I checked some of my dev HTTP requests in postman and they only use ARM ID, but that stuff circumvents ARM, so I'm not sure it's correct, especially for system identity (which is tricky in dev env)
> 
> Comparing to how the azurerm terraform provider solves this exact problem I believe you are correct and this feature is also designed to be receive multiple user assigned IDs. Please check out the following code:
> 
> https://github.com/hashicorp/terraform-provider-azurerm/blob/8bc0e5637346eeea808f5b63b97eed69835005b5/internal/services/containers/kubernetes_cluster_resource.go#L3265-L3283
> 
> I think a better API would be to reuse the idea in `AzureMachinePool` with the [UserAssignedIdentities](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1beta1/azuremachinepool_types.go#L144C6-L149) list on `AzureManagedControlPlane` and do a loop that looks like the one in the tf provider to add empty `ManagedClusterIdentityUserAssignedIdentitiesValue` with the ID as the map key.

So it looks like AKS only accepts a single user-assigned ID (which makes sense):

E0620 15:58:58.430110       1 controller.go:329]  "msg"="Reconciler error" "error"="error creating AzureManagedControlPlane capz-e2e-n5ol1g/capz-e2e-n5ol1g-aks: failed to reconcile AzureManagedControlPlane service managedcluster: failed to create resource capz-e2e-n5ol1g-aks/capz-e2e-n5ol1g-aks (service: managedcluster): Code=\"BadRequest\" Message=\"The count of user-assigned identity specified in identity.userAssignedIdentities is 2, while expected value is 1.\"" "AzureManagedControlPlane"={"name":"capz-e2e-n5ol1g-aks","namespace":"capz-e2e-n5ol1g"} "controller"="azuremanagedcontrolplane" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedControlPlane" "name"="capz-e2e-n5ol1g-aks" "namespace"="capz-e2e-n5ol1g" "reconcileID"="51a30f2d-bd9d-40ab-af2a-dec8b1ca4f9e"

I'll go back to using a single string interface in the CAPZ property definition so users aren't confused.

File: docs/book/src/topics/managedcluster.md
Comment (jackfrancis) [inline docs/book/src/topics/managedcluster.md:288]: @nojnhuh with your recent streamlining of AKS feature docs is there a better way to document this?

Comment (nojnhuh) [reply docs/book/src/topics/managedcluster.md:288]: I tried to establish some amount of a pattern in #3550 for what to include in the godoc comments for the type definitions. My initial sense is that what's included here overlaps with the godoc comments enough to justify not needing a separate section here. Definitely no hard rules though, so feel free to add anything here you think is valuable. I'd at least include the AKS doc link in the godoc in any case.

Comment (jackfrancis) [reply docs/book/src/topics/managedcluster.md:288]: godocs are generated after merge to main?

Comment (nojnhuh) [reply docs/book/src/topics/managedcluster.md:288]: The deploy preview for this PR has the generated version of these changes: https://deploy-preview-3425--kubernetes-sigs-cluster-api-provider-azure.netlify.app/reference/v1beta1-api.html#infrastructure.cluster.x-k8s.io/v1beta1.IdentityProfile

Comment (jackfrancis) [reply docs/book/src/topics/managedcluster.md:288]: Love it, will remove the add'l matter I added to the markdown.

File: azure/services/managedclusters/spec.go
Comment (nojnhuh) [inline azure/services/managedclusters/spec.go:597]: nit: removing most of this level of indentation might make this a bit more readable:
```suggestion
	if identityProfile.Type == "" {
		return
	}
	managedClusterIdentity = &containerservice.ManagedClusterIdentity{
```

File: azure/services/managedclusters/spec.go
Comment (nojnhuh) [inline azure/services/managedclusters/spec.go:606]: nit: inlining this might be slightly less verbose? At the expense of some possible memory optimization Go does with `make(..., 1)` that doesn't happen with a literal.
```suggestion
			managedClusterIdentity.UserAssignedIdentities = map[string]*containerservice.ManagedClusterIdentityUserAssignedIdentitiesValue{
				identityProfile.UserAssignedIdentityResourceID: {},
			}
```

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (nawazkh) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:698]: ^ looks like the implemented error is supposed to be shown to the user when `m.Spec.IdentityProfile.Type == ManagedControlPlaneIdentityTypeUserAssigned` and  `m.Spec.IdentityProfile.UserAssignedIdentityResourceID != ""` .
So "_`should`_ be empty if IdentityProfile.Type is SystemAssigned" is about right?

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (nawazkh) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:707]: I agree with Jon that we could return `allErrs`, but is there a reason for aggregating the errors which I may be missing?

File: azure/services/managedclusters/spec.go
Comment (nawazkh) [reply azure/services/managedclusters/spec.go:597]: ^ in addition to the above change, we might want to add an error as well.

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:698]: Ah yes, I can't read 🙃

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:707]: If you mean "why do we return multiple errors" then in general, we want to return as much information as possible to the user for each request so they get one "spec.A is invalid, spec.B is invalid, ..." error on the first request vs. "spec.A is invalid." Fix spec.A. "spec.B is invalid." Fix spec.B...

If you mean "why are we calling `kerrors.NewAggregate` then that's because that's one way of turning a `[]error` into a single `error`.

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:63]: Good question. The current implementation simply hard-codes the AKS cluster config to use system-assigned identity. The scope of this is to add user-assigned, so I hadn't considered including backfilling enabling service principal config.

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:442]: love

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:691]: I can test

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:698]: fun with natural language inversions!

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:707]: I think I'd have to refactor more to use the simpler approach, as the `Validate` method is doing the error wrapping by enumerating through a bunch of validator funcs (`validateIdentityProfile` is a new one I added here) and each of those funcs is returning an `err`.

Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:707]: Ah I missed that, this is definitely fins as-is then IMO.

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_types.go:63]: This would be easy and safe enough to add later if it's not a simple oversight.

File: azure/services/managedclusters/spec.go
Comment (jackfrancis) [reply azure/services/managedclusters/spec.go:602]: I initially excluded the error check but then when I went to write the UT I thought "do we really want to rely upon the webhook forever to make this function sane?" and decided including an error response as a low-maintenance, high-defense outcome. So I think I still prefer it.

File: azure/services/managedclusters/spec.go
Comment (jackfrancis) [reply azure/services/managedclusters/spec.go:597]: we are returning an error it's just that this func is using named response variables so things are maybe a little cute (but using a named `*containerservice.ManagedClusterIdentity` response variable in particular does clean up the code a bit)

File: azure/services/managedclusters/spec.go
Comment (jackfrancis) [reply azure/services/managedclusters/spec.go:606]: deal

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:181]: I don't love naming the struct just `identity` as that sort of suggests a shorthand property key for the type of identity.

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:63]: Yeah I think it should be something we add in reponse to user demand (system-assigned identity is definitely the happy path IMO)

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_types.go:181]: If we decide we need to expose `properties.identityProfile` is there a reasonable name we could use for it in the CAPZ API?

Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:181]: Actually I was misunderstood your initial comment. You're right, this really is a sort of worst-case scenario in terms of API semantics. :(

From a test cluster built from this branch and configured for control plane user-assigned identity:

```
...
        "identityProfile": {
            "kubeletidentity": {
                "resourceId": "/subscriptions/<redacted>/resourcegroups/MC_capz-e2e-hikyno-aks_capz-e2e-hikyno-aks_eastus/providers/Microsoft.ManagedIdentity/userAssignedIdentities/capz-e2e-hikyno-aks-agentpool",
                "clientId": "<redacted>",
                "objectId": "<redacted>"
            }
        },
    "identity": {
        "type": "UserAssigned",
        "userAssignedIdentities": {
            "/subscriptions/<redacted>/resourcegroups/capz-ci/providers/Microsoft.ManagedIdentity/userAssignedIdentities/umd-test": {
                "clientId": "<redacted>",
                "principalId": "<redacted>"
            }
        }
    },
...
```

The latter is the one that we're using w/ the new CAPZ interface here.

Would it be easier to implement the kubelet `identityProfile` in this PR as well so we ensure we do them both right w/ respect to one another? (I don't think there's that much more work.)

In either case yeah we probably wanna rename to `identity` for the control plane identity interface.

Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_types.go:181]: I'm not sure I fully understand the differences between those fields. Do those actually overlap at all and where it's CAPZ's responsibility to keep them in sync?

Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:181]: They're the first two in the list here:

- https://learn.microsoft.com/en-us/azure/aks/use-managed-identity#summary-of-managed-identities

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go]: The AKS is accepting changes to the fields. Though the transition between, e.g., changing from one user-assigned identity resource to another, seemed very slow (cluster was in an Updating state for a long time).

For now I think we should keep these fields mutable (no webhook violations), but hold off on adding E2E tests that validate these updates in practice until we can get a better handle on how reliable these operations are in practice.

File: azure/services/managedclusters/spec.go
Comment (jackfrancis) [inline azure/services/managedclusters/spec.go:370]: This was done to appease the cyclomatic complexity gods.

**Comment by jackfrancis** [comment]
/assign @mboersma 

Matt, are you able to review and approve this so that it lands in the v1.10 milestone? This ticks off two features tracked by @dtzar in #3446

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (mboersma) [inline api/v1beta1/azuremanagedcontrolplane_types.go:55]:
```suggestion
	// ManagedControlPlaneIdentityTypeSystemAssigned Use an implicitly created system-assigned managed identity to manage
```

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (mboersma) [inline api/v1beta1/azuremanagedcontrolplane_types.go:57]:
```suggestion
	// system-assigned managed identity to manipulate Azure resources.
```

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (mboersma) [inline api/v1beta1/azuremanagedcontrolplane_types.go:60]:
```suggestion
	// Components in the control plane such as kube-controller-manager will use the specified user-assigned
```

**Review by mboersma — COMMENTED** [review]
Looks good to moi, just had some minor documentation comments.

**Comment by mboersma** [comment]
/test pull-cluster-api-provider-azure-apidiff

`Pod could not be created...` prow failure.

**Comment by dtzar** [comment]
Assuming the provider azure test passes, are we good to merge this @mboersma ?

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Comment by maciaszczykm** [comment]
@jackfrancis This change seems to break reconcile process for me. Control plane sticks in updating status, after debug found out that diff (azure/services/managedclusters/spec.go:415) is not empty:

```diff
  &containerservice.ManagedCluster{
  	... // 2 identical fields
  	ExtendedLocation: nil,
  	Identity:         &{Type: "SystemAssigned"},
  	ManagedClusterProperties: &containerservice.ManagedClusterProperties{
  		... // 22 identical fields
  		APIServerAccessProfile: nil,
  		DiskEncryptionSetID:    nil,
- 		IdentityProfile:        nil,
+ 		IdentityProfile: map[string]*containerservice.UserAssignedIdentity{
+ 			"kubeletidentity": &{
+ 				ResourceID: &"/subscriptions/[REMOVED]/resourcegrou"...,
+ 			},
+ 		},
  		PrivateLinkResources: nil,
  		DisableLocalAccounts: nil,
  		... // 3 identical fields
  	},
  	Tags:     nil,
  	Location: nil,
  	... // 4 identical fields
  }
```

I did not specify this field. It looks like it was filled in on the AKS side but not in control plane resource.

**Comment by dtzar** [comment]
@maciaszczykm can you please file an issue for this?

**Comment by nojnhuh** [comment]
PR open to fix: #3682

**Pull request merged**
